### ISSUE TITLE
Eval torch compile slower than eager for Torch 2.1.0

### ISSUE THREAD

**Issue by ajindal1** [pytorch/pytorch#110666]
### 🐛 Describe the bug

Getting significantly higher eval time for torch compile compared to eager mode. Train time is close to expected. The speedup for eval mode is around 0.017, while it should be greater than 1. Complete logs are given below. I am using Pytorch's 2.1.0 image (pytorch/pytorch:2.1.0-cuda11.8-cudnn8-devel) and Pytorch torch compile [tutorial](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) for this.

```
eager: 0.2973194274902344
compile: 63.84834765625
eager eval time 0: 0.03339263916015625
eager eval time 1: 0.026991615295410155
eager eval time 2: 0.027637760162353517
eager eval time 3: 0.027448320388793947
eager eval time 4: 0.02708684730529785
eager eval time 5: 0.02836787223815918
eager eval time 6: 0.0277708797454834
eager eval time 7: 0.027261951446533202
eager eval time 8: 0.027030527114868166
eager eval time 9: 0.03097599983215332
~~~~~~~~~~
compile eval time 0: 0.532326416015625
compile eval time 1: 0.6827396850585937
/opt/conda/lib/python3.10/site-packages/torch/_inductor/cudagraph_trees.py:2049: UserWarning: Unable to hit fast path of CUDAGraphs because of pending, uninvoked backwards. Consider running with torch.no_grad() or using torch._inductor.cudagraph_mark_step_begin() before each model invocation
  warnings.warn(
compile eval time 2: 0.9548626098632812
compile eval time 3: 1.202206787109375
compile eval time 4: 1.483588623046875
compile eval time 5: 1.758994384765625
compile eval time 6: 1.985332275390625
compile eval time 7: 2.296648681640625
compile eval time 8: 2.576333740234375
compile eval time 9: 2.781447265625
~~~~~~~~~~
(eval) eager median: 0.027543040275573732, compile median: 1.62129150390625, speedup: 0.01698833319561168x
~~~~~~~~~~
eager train time 0: 0.2526566009521484
eager train time 1: 0.07625110626220703
eager train time 2: 0.07144239807128906
eager train time 3: 0.0705730209350586
eager train time 4: 0.07013785552978516
eager train time 5: 0.07054541015625
eager train time 6: 0.07060479736328125
eager train time 7: 0.07041126251220703
eager train time 8: 0.0709591064453125
eager train time 9: 0.07169843292236328
~~~~~~~~~~
skipping cudagraphs due to input mutation
compile train time 0: 365.358375
compile train time 1: 3.111400390625
compile train time 2: 0.045563873291015626
compile train time 3: 0.033967136383056644
compile train time 4: 0.03476684951782227
compile train time 5: 0.035434497833251956
compile train time 6: 0.03588710403442383
compile train time 7: 0.039874561309814455
compile train time 8: 0.03737395095825195
compile train time 9: 0.035451904296875
~~~~~~~~~~
(train) eager median: 0.07078195190429687, compile median: 0.03663052749633789, speedup: 1.9323213926245872x
~~~~~~~~~~

```

However, it works fine with torch 2.0.1 image (pytorch/pytorch:2.0.1-cuda11.7-cudnn8-devel). 

```
compile eval time 9: 0.010388480186462403
~~~~~~~~~~
(eval) eager median: 0.025155088424682618, compile median: 0.011081727981567382, speedup: 2.2699608279975774x
~~~~~~~~~~

compile train time 8: 0.03183718490600586
compile train time 9: 0.03175424003601074
~~~~~~~~~~
(train) eager median: 0.07355648040771484, compile median: 0.033122287750244145, speedup: 2.220754827153286x
~~~~~~~~~~

```

### Versions

Collecting environment information...
PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-1047-azure-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB
GPU 2: Tesla V100-PCIE-16GB
GPU 3: Tesla V100-PCIE-16GB

Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             24
On-line CPU(s) list:                0-23
Thread(s) per core:                 1
Core(s) per socket:                 12
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              79
Model name:                         Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz
Stepping:                           1
CPU MHz:                            2593.994
BogoMIPS:                           5187.98
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          768 KiB
L1i cache:                          768 KiB
L2 cache:                           6 MiB
L3 cache:                           70 MiB
NUMA node0 CPU(s):                  0-11
NUMA node1 CPU(s):                  12-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT Host state unknown
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt md_clear

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.1.0
[pip3] torchaudio==2.1.0
[pip3] torchelastic==0.2.2
[pip3] torchvision==0.16.0
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46343
[conda] mkl-service               2.4.0           py310h5eee18b_1
[conda] mkl_fft                   1.3.8           py310h5eee18b_0
[conda] mkl_random                1.2.4           py310hdb19cb5_0
[conda] numpy                     1.26.0          py310h5f9d8c6_0
[conda] numpy-base                1.26.0          py310hb5e798b_0
[conda] pytorch                   2.1.0           py3.10_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.1.0               py310_cu118    pytorch
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchtriton               2.1.0                     py310    pytorch
[conda] torchvision               0.16.0              py310_cu118    pytorch

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by benureau** [comment]
The tutorial html might also need to be updated after this fixed. The sentence "And indeed, we can see that running our model with torch.compile results in a significant speedup." is not consistent with the output just above on the documentation page.

**Comment by gchanan** [comment]
CC @williamwen42

**Comment by malfet** [comment]
@williamwen42 I understand the fix to the tutorial, but can you please help me understand, did it show the speedup in 2.0, or because it was comparing nograd for eager vs forward+backward for compile it was always slower? (if this is the case, should we just move this issue to tutorials repo, as there is nothing wrong with `torch.compile`?)

**Comment by ezyang** [comment]
Agreed with @malfet, isn't the real bug torch.no_grad stopped working?

**Comment by voznesenskym** [comment]
@williamwen42 why did you close this?

**Comment by williamwen42** [comment]
It was an auto-close - reopening.

**Comment by bdhirsh** [comment]
Triage results:

When I run the repro, I see this warning:
```
UserWarning: Unable to hit fast path of CUDAGraphs because of pending, uninvoked backwards. Consider running with torch.no_grad() or using torch._inductor.cudagraph_mark_step_begin() before each model invocation
```

There are a few things going on here:

(1) When you compile the code below, AOTAutograd will decide that it is a **training** graph, and not an inference graph. This is not a regression and has always been the case:
```
@torch.compile
def generate(x):
    with torch.no_grad():
        # assume model has some params with requires_grad=True
        return model(*x)
```
Why does AOTAutograd make this a training graph? The code for this lives [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L4124): AOTAutograd inspects the global autograd state at the start of compiling, but the no_grad() call lives inside the compiled graph, and won't flip the state until we start tracing.

(2) According to @eellison, we can no longer easily cudagraph-ify this code now that cudagraph trees are used by default. Importantly, torch.compile **previously** used a separate memory pool for every compilation (making case safe, but causing other cases to use excessive memory), but cudagraph trees allow sharing of memory pools across compiled regions, so we can't easily just blindly turn on cudagraphs when we have a pending backwards graph. In this particular case the backward graph is completely empty, but it's difficult to tell inductor this during the forward without some larger changes between aot autograd and inductor.

Given all that and given that we can't reasonably revert cudagraph trees, the simplest fix is probably to forward-fix AOTAutograd to be smarter and recognize that the graph above is an inference graph. This will probably involve inspecting the torch graph, and figuring out that every op inside of it was effectively under a no_grad.

**Comment by atalman** [comment]
hi @bdhirsh could you please cherry pick fix for this issue to 2.1.1 milestones ?
